Start plugin

tcanabrava opened the pull request:
implementing the first draft of the plugin system.

tcanabrava committed:
Added the beginning of the work for firmware plugins

Added the beginning of the work for firmware plugins,
the AtCore class should handle the plugin loading and
discover the low level stuff, managing the other classes.

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

tcanabrava committed:
Try to load the plugin if firmware found.

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

tcanabrava committed:
Reorganize the Find Firmware method.

Now the find firmware doesn't calls a lambda (as the code
grew quite big and I'm a big fan of not having a ladder-code)

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

tcanabrava committed:
Change class to struct.

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

tcanabrava committed:
Added first functional (untested) plugin

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

sithlord48 commented on src/atcore.cpp:
```diff
@@ -0,0 +1,91 @@
+        QCoreApplication::processEvents( QEventLoop::AllEvents, 100 );
+    }
+
+    // What this command does?
```
M115 makes the firmware return a fw id string. along with some other data about the printer/fw.

tcanabrava committed:
Added a new plugin for testing purposes.

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

tcanabrava committed:
Finished first implementation of a plugin system.

Ok, the code currently is a bit of a mess - the AtCore clearly
lacks a good architecture and we must stop what we are doing now
to think a bit about the architecture that we want to use.

Currently too much of setup is done within the test application,
this should be used inside of the AtCore - a init method perhaps.

I don't know what should be inside the firmwares, so one with
more knowledge of printing functions should add what's missing there

Right now there's a Protocol Layer / Firmware Plugin that are
side by side in the d-pointer, and after a brief talk with Patrick
that's ok, but the plugin should translate what will be send to the
ProtocolLayer and vice versa - that's what I don't have any idea
on how to do.

Signed-off-by: Tomaz Canabrava <tomaz.canabrava@gmail.com>

patrickelectric commented:
About grbl, add this branch to use M115 command (In the commit message or in source)
https://github.com/patrickelectric/grbl/tree/add_m115

patrickelectric commented:
@tcanabrava please close the PR.
